pkg/cmd/experimental/syncgroups/cli/prunegroups.go

```diff
@@ -0,0 +1,237 @@
+	GroupInterface osclient.GroupInterface
+
+	// Stderr is the writer to write warnings and errors to
+	Stderr io.Writer
+
+	// Out is the writer to write output to
+	Out io.Writer
+}
+
+func NewPruneGroupsOptions() *PruneGroupsOptions {
+	return &PruneGroupsOptions{
+		Stderr:    os.Stderr,
+		Whitelist: []string{},
+	}
+}
+
+func NewCmdPruneGroups(name, fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	options := NewPruneGroupsOptions()
+	options.Out = out
+
+	whitelistFile := ""
+	blacklistFile := ""
+	configFile := ""
+
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s [WHITELIST] [--whitelist=WHITELIST-FILE] [--blacklist=BLACKLIST-FILE] --sync-config=CONFIG-SOURCE", name),
+		Short:   "Prune OpenShift groups referencing missing records on an external provider.",
+		Long:    pruneGroupsLong,
+		Example: fmt.Sprintf(pruneGroupsExamples, fullName),
+		Run: func(c *cobra.Command, args []string) {
+			if err := options.Complete(whitelistFile, blacklistFile, configFile, args, f); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(c, err.Error()))
+			}
+
+			if err := options.Validate(); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(c, err.Error()))
+			}
+
+			err := options.Run(c, f)
+			if err != nil {
+				if aggregate, ok := err.(kerrs.Aggregate); ok {
+					for _, err := range aggregate.Errors() {
+						fmt.Printf("%s\n", err)
+					}
+					os.Exit(1)
+				}
+			}
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	cmd.Flags().StringVar(&whitelistFile, "whitelist", whitelistFile, "path to the group whitelist file")
+	cmd.Flags().StringVar(&blacklistFile, "blacklist", whitelistFile, "path to the group blacklist file")
+	// TODO enable this we're able to support string slice elements that have commas
+	// cmd.Flags().StringSliceVar(&options.Blacklist, "blacklist-group", options.Blacklist, "group to blacklist")
+	cmd.Flags().StringVar(&configFile, "sync-config", configFile, "path to the sync config")
+	cmd.Flags().BoolVar(&options.Confirm, "confirm", false, "if true, modify OpenShift groups; if false, display groups")
+	cmdutil.AddPrinterFlags(cmd)
+	cmd.Flags().Lookup("output").DefValue = "yaml"
+	cmd.Flags().Lookup("output").Value.Set("yaml")
+
+	return cmd
+}
+
+func (o *PruneGroupsOptions) Complete(whitelistFile, blacklistFile, configFile string, args []string, f *clientcmd.Factory) error {
+	whitelist, err := buildNameList(GroupSyncSourceOpenShift, args, whitelistFile)
+	if err != nil {
+		return err
+	}
+	o.Whitelist = whitelist
+	blacklist, err := buildNameList(GroupSyncSourceOpenShift, []string{}, blacklistFile)
+	if err != nil {
+		return err
+	}
+	o.Blacklist = blacklist
+
+	config, err := decodeSyncConfigFromFile(configFile)
+	if err != nil {
+		return err
+	}
+	o.Config = *config
+
+	osClient, _, err := f.Clients()
+	if err != nil {
+		return err
+	}
+	o.GroupInterface = osClient.Groups()
+
+	return nil
+}
+
+func (o *PruneGroupsOptions) Validate() error {
```

Comment:
> Why do we need to validate that? Especially now that the addition of the interface is not conditional on anything? If we fail to do it, Complete will error.

Delegation.  If someone builds one of these, they might mess it up.  `Validate` will smack them.